Show warnings in pytest and turn them into errors

pllim opened the pull request:
**DO NOT MERGE**

We are ignoring warnings right now but as @bsipocz stated, they should be cleaned up at some point. This PR show them for the purpose of clean-up campaign.

* `convolution` -- #7983
* `coordinates` -- #7989 and #8033
* `cosmology` -- #7992
* `io.ascii` -- #7997
* `io.fits` -- #7998 (partial only), #8045 and #8749 by @saimn 
* `io.votable` -- #8715 by @astrofrog and #8756
* `modeling` -- #9538 
* `time` -- #8754
* `timeseries.periodograms` -- #8964 
* `units` -- #8755 by @mhvk 
* `utils` --  #8034
* `wcs` -- #9010
* `visualization` -- #8001 and #8007 by @astrofrog 
* Warnings from `doctestplus` -- astropy/pytest-doctestplus#59
* Sweep through an updated log to see if anything new pops up

Update: Over the months, I might have lost track of some PRs that got merge that got rid of the warnings. I apologize if I missed your PR(s). Please update the list above accordingly if you so desired. See the xref PRs in the comments thread below.

Fun fact: #6163

pllim commented:
Example log: https://travis-ci.org/astropy/astropy/jobs/443714275

Whee, only 2089 warnings... 😅

bsipocz commented:
> Whee, only 2089 warnings...

well, there was good reasons for adding the ignorance to the config ;)

pllim commented:
Adding a "sprint" tag in case people feel like sprinting for this one, one subpackage at a time.

astrofrog commented:
The warnings about figures not being closed by Matplotlib can be fixed by making sure that pytest-mpl is installed (even if it's not actually enabled with ``--mpl``)

astrofrog commented:
@pllim - can you install pytest-mpl for the job where you are checking the warnings? I think we might want to just add it as one of the packages that gets installed with pytest-astropy so it's always there by default.

pllim commented:
re: pytest-mpl -- Didn't it used to be in `pytest-astropy` but then taken out due to some problems? 🤔

pllim commented:
@astrofrog -- The build with `pytest-mpl` https://travis-ci.org/astropy/astropy/jobs/447178372

bsipocz commented:
hmm, it's worse than when we've started?

pllim commented:
😱 I hope I didn't accidentally make things worse in my attempts to fix them. 😱

pllim commented:
Ah, but before we go into full panic. The earlier number was from Python 3.5 run, while the latter one from Python 3.7. If I go back to that earlier log and pick the Python 3.7 build (https://travis-ci.org/astropy/astropy/jobs/443714273), there were actually 2689 warnings. So now we have 541 less. 😬

drdavella commented:
I need to refresh myself on exactly how this works, but I wonder how many of these warnings come from outside astropy? If you enable warnings using the Python interpreter outside of pytest, you end up getting warnings from all kinds of libraries that we don't really care about.

pllim commented:
If I try to comb through all 2k plus warnings to answer your question, my eyes would bleed. I think this one is here to stay for a while. It is more manageable when we approach this little by little until maybe 100 left and then we can be aggressive about it. :see_no_evil:

astrofrog commented:
Just for info, it seems that potentially as many as 1000 warnings come from io.votable. Of the remaining ones, here are the ones that appear more than 10 times:

```
  195 ErfaWarning: ERFA function "utctai" yielded 5 of "dubious year (Note 3)"
  160 AstropyWarning: failed to download http://maia.usno.navy.mil/ser7/finals2000A.all, using local IERS-B: <urlopen error An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: maia.usno.navy.mil>
  103 RuntimeWarning: divide by zero encountered in true_divide
   94 AstropyUserWarning: The convolve_fft version of boundary=None is equivalent to the convolve boundary='fill'.  There is no FFT equivalent to convolve's zero-if-kernel-leaves-boundary
   80 RuntimeWarning: divide by zero encountered in log
   70 RuntimeWarning: invalid value encountered in true_divide
   47 ErfaWarning: ERFA function "utctai" yielded 2 of "dubious year (Note 3)"
   44 AstropyUserWarning: Model is linear in parameters; consider using linear fitting methods.
   32 RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
   24 RuntimeWarning: divide by zero encountered in log1p
   19 FITSFixedWarning: The WCS transformation has more axes (2) than the image it is associated with (0)
   18 AstropyUserWarning: The following header keyword is invalid or follows an unrecognized non-standard convention:
   15 ErfaWarning: ERFA function "utctai" yielded 3 of "dubious year (Note 3)"
   14 AstropyUserWarning: The fit may be unsuccessful; check fit_info['message'] for more information.
   14 AstropyUserWarning: The fit may be poorly conditioned
   13 UserWarning: Warning: converting a masked element to nan.
```

pllim commented:
Interesting... I thought @eteq got rid of all the "dubious year" warnings from ERFA but maybe not!

pllim commented:
> as many as 1000 warnings come from io.votable

I won't be surprised... :see_no_evil:

bsipocz commented:
`io.votable` clearly needs someone who loves and cherishes it. It's quite sad actually that while vo is_a_thing ™️  yet no one steps up to maintain it.

bsipocz commented:
I've naively rebased to see how we stand. Well, these warnings are not going away on their own accord, there are 2697 left :(

pllim commented:
> these warnings are not going away on their own accord

Nope. In fact, a few more might have crept in since...

astrofrog commented:
The latest version of pytest groups warnings into where they originated so that might help figure out some of the most common warnings.

astrofrog commented:
Here's the latest log for anyone interested in helping: https://gist.github.com/astrofrog/bc2197c3f11324b84d8a62d252ed4dbd

bsipocz commented:
I've started to clean up the convolution ones yesterday, but didn't get too far an instead kept on triaging. Cleaning up these are very much kills your soul...

pllim commented:
Hey, hey... With #8715 merged, number of warnings went down to 1752 (https://travis-ci.org/astropy/astropy/jobs/536364376). As far as `votable` is concerned, it still emits some warnings, which are not controlled by `verify` (or `pedantic`). For those, we can insert `warnings.simplefilter()` on a case by case basis for the affected tests.

astrofrog commented:
>  For those, we can insert warnings.simplefilter() on a case by case basis for the affected tests.

Or better, ``pytest.warns`` :)

pllim commented:
Hmm, weird, Appveyor is failing because #5797 is seemingly resolved but only for this PR, not on `master`. Note to self: Follow up on this when this PR is ready for final review.

pllim commented:
https://travis-ci.org/astropy/astropy/jobs/559217591 shows 881 warnings now (not as optimistic as the 550 that Brigitta saw above) but less than 1000!

bsipocz commented:
We'll get there eventually, maybe even for 4.0.
I might not have had some of the optional dependencies installed thus the lower number. 🤷‍♀

mhvk commented:
I'm getting rid of the warnings from units - see #9005

bsipocz commented:
Locally it's getting worse for me, not even sure why all those modeling tests fail while they are fine on CI. Anyway, bottom line is the number 910:
```
=================== 51 failed, 12954 passed, 338 skipped, 64 xfailed, 910 warnings in 205.84 seconds ===================
```

pllim commented:
@bsipocz , as we merge new PRs, new warnings might creep in. Can't guarantee it always goes down with latest `master`. :grimacing:

bsipocz commented:
@pllim - sure. I also have most things that matter on their dev branch locally, that might cause some extra warnings, too.

pllim commented:
Argh, climbed back up to 1290 warnings. :see_no_evil:

pllim commented:
Down to 177 warnings (not counting remote data tests). 🎉 🤞

mhvk commented:
Hmm, one of the ones in `Time` is new - apparently at the end of October with commit 1f8faf722e55e02e4d30025d4746e05bce5c6343 we finally started to support star dates using the new `TimeStardate` format :fireworks: :fireworks: (though formal process seems to have been bypassed with a direct merge to master by Captain @astrofrog...) - and the tests raise warnings because stardate is out of range for UTC. Indeed, I could find no mention of timescale in the [reference given](http://trekguide.com/Stardates.htm#TNG), but I would *think* that since their day is not even equal to an earth day, there is little need for the leap seconds in UTC and it is more likely atomic time is used, i.e., TAI or perhaps TDB.

p.s. This does show that we should try to get rid of the warnings and then merge this PR - it becomes a bit whack-a-klingon otherwise.

mhvk commented:
for `time`, see #9563; I think `units` is still clean.

pllim commented:
> then merge this PR

I wish... 150+ warnings are a lot less than before but they are still spread across several subpackages (`io.ascii`, `io.misc`, `utils`, `table`, to name a few) and will take some time to wrap up.

If I can help it, I will even set `setup.cfg` to make `pytest` treat warnings as errors though that might be too ambitious?

mhvk commented:
I think we should end up with treating warnings as errors - now caught several bugs with this already!

astrofrog commented:
Latest count, without remote tests:

```
astropy/config: 0
astropy/constants: 0
astropy/convolution: 0
astropy/coordinates: 11
astropy/cosmology: 0
astropy/io: 71
astropy/modeling: 0
astropy/nddata: 0
astropy/samp: 0
astropy/stats: 6
astropy/table: 30
astropy/tests: 2
astropy/time: 0
astropy/timeseries: 1
astropy/uncertainty: 0
astropy/units: 0
astropy/utils: 18
astropy/visualization: 2
astropy/wcs: 0
```

pllim commented on astropy/conftest.py:
```diff
@@ -27,12 +27,14 @@
-    modules_to_ignore_on_import=['requests'],
+    # It also ignores collections.abc warning from html5lib that has been fixed
+    # in dev but not in release (html5lib/html5lib-python#419)
+    modules_to_ignore_on_import=['requests', 'html5lib'],
```
Note to self: Change `html5lib` to `bleach`. Turns out `bleach` bundles `html5lib` code. Sigh.

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,14 @@ doctest_plus = enabled
+filterwarnings =
+    ignore:unclosed file:ResourceWarning
+    ignore:unclosed <socket:ResourceWarning
+    ignore:numpy.ufunc size changed:RuntimeWarning
```
Note to self: Need to add another line to ignore this one, as I don't know how to silence it otherwise:
```
astropy/utils/timer.py:177
  .../astropy/utils/timer.py:177: AstropyDeprecationWarning:
  The timefunc function is deprecated and may be removed in a future version.
          Use astroquery.utils.timer.timefunc instead.
    @timefunc(num_tries=1, verbose=False)
```

This is because I am decorating a method with deprecated decorator. But both the method's class and the decorator are deprecated and I purposely excluded them from API doc collection, etc (you can check it on `master`). Still this warning pops up. 🤷‍♀

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,14 @@ doctest_plus = enabled
 qt_no_exception_capture = 1
+filterwarnings =
+    ignore:unclosed file:ResourceWarning
+    ignore:unclosed <socket:ResourceWarning
```
```suggestion
    ignore:unclosed <socket:ResourceWarning
    ignore:unclosed <ssl.SSLSocket:ResourceWarning
```

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,16 @@ doctest_plus = enabled
+    ignore:unclosed <socket:ResourceWarning
+    ignore:unclosed <ssl.SSLSocket:ResourceWarning
+    ignore:The timefunc function is deprecated
+    ignore:numpy.ufunc size changed:RuntimeWarning
```
@mhvk , should we worry about `numpy.ufunc size changed` or can we ignore it? I wonder if it is a complain that is exclusive to `conda`.

astrofrog replied:
I think we should indeed ignore this, it can happen with pip too in some cases, but it's beyond the scope of our test suite.

pllim replied:
Do you know why `numpy.ufunc` is tormenting us so? I am curious.

astrofrog replied:
Not fully, but it's a common issue with some combinations of numpy and other packages, it's common enough we should just always ignore it.

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,16 @@ doctest_plus = enabled
 xfail_strict = true
 qt_no_exception_capture = 1
+filterwarnings =
+    ignore:unclosed file:ResourceWarning
```
The `ResourceWarning` rabbit hole is too deep. We can open an issue to follow-up after this is merged, thought I fear that issue may remain open for a long time yet.

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,16 @@ doctest_plus = enabled
 asdf_schema_root = astropy/io/misc/asdf/data/schemas
 xfail_strict = true
 qt_no_exception_capture = 1
+filterwarnings =
```
When we think we have addressed all the remaining warnings, 🔥 
```suggestion
filterwarnings =
    error
```

pllim commented on astropy/conftest.py:
```diff
@@ -27,12 +27,14 @@
     # the `requests` module. It only appears when both asdf and sphinx are
     # installed. This can be removed once pyopenssl 1.7.20+ is released.
-    modules_to_ignore_on_import=['requests'],
+    # It also ignores collections.abc warning from html5lib that has been fixed
```
```suggestion
    # It also ignores collections.abc warning from html5lib in bleach that has been fixed
```

pllim commented on astropy/io/fits/tests/test_hdulist.py:
```diff
@@ -620,8 +619,7 @@ def test_update_with_truncated_header(self):
         with fits.open(self.temp('temp.fits')) as hdul:
             assert (hdul[0].data == data).all()
 
-    @pytest.mark.xfail(platform.system() == 'Windows',
```
I have no idea why two tests in FITS marked as xfail on Windows would start passing for this PR, causing Windows job to fail. 🤷‍♀

pllim commented on astropy/utils/tests/test_data.py:
```diff
@@ -1000,12 +1000,15 @@ def test_is_url_in_cache_local(temp_cache, valid_urls, invalid_urls):
     assert not is_url_in_cache(nonexistent)
 
 
+@pytest.mark.filterwarnings('ignore:Not clearing data cache - '
```
@aarchiba , if you are okay with it, we can ignore this warning for now and then open a follow-up issue to deal with it post-4.0?

aarchiba replied:
How tight are we for time? I can probably get this straightened out in a day or two but I don't want to delay anything.

pllim replied:
Not *that* tight! 😅  Thanks!

aarchiba replied:
I approve of the quest to banish all warnings! PINT emits so many warnings users can't reliably tell when they're losing accuracy.

aarchiba replied:
I think I fixed this one.

pllim replied:
Yup. Thanks!

pllim commented:
Only two left and they are pending merging of #9593 and #9594.

pllim commented:
"This branch has conflicts that must be resolved" -- I don't understand... I just rebased and I find no conflict.

pllim commented on docs/timeseries/index.rst:
```diff
@@ -252,8 +252,11 @@ Let's take a look at the folded time series:
     >>> from astropy.stats import sigma_clipped_stats
-    >>> mean, median, stddev = sigma_clipped_stats(ts_folded['sap_flux'])  # doctest: +REMOTE_DATA
+    >>> with warnings.catch_warnings():
+    ...     warnings.simplefilter('ignore')  # Ignore warning about invalid values
```
I guess we can use the new `IGNORE_WARNINGS` when a new version of `pytest-doctestplus` is out.

astrofrog replied:
This can now be done in principle!

pllim commented on astropy/utils/iers/tests/test_iers.py:
```diff
@@ -302,11 +302,13 @@ def test_simple(self):
             # Now the time range should be different.
             assert dat['MJD'][0] == 57359.0 * u.d
-            assert dat['MJD'][-1] == (57539.0 + 60) * u.d
+            assert dat['MJD'][-1] == (57539.0 + 48) * u.d
```
This looks sketchy but is out of scope for this PR. I just wanted the test to pass. See #9600.

astrofrog replied:
For the purposes of merging this PR I'd be more comfortable with just ignoring warnings from this test rather than changing the value here?

pllim replied:
This is not a warning; it is outright an `AssertionError`. Should I just let it fail until @aarchiba has a chance to investigate #9600 or should I `xfail` it?

astrofrog replied:
I feel like we should just xfail it for now unless #9600 is addressed before we need to merge this PR.

pllim commented on astropy/table/tests/test_mixin.py:
```diff
@@ -154,8 +154,15 @@ def test_io_time_write_fits_standard(tmpdir, table_types):
     # Show that FITS format succeeds
-    t.write(filename, format='fits', overwrite=True)
-    tm = table_types.read(filename, format='fits', astropy_native=True)
+    with pytest.warns(
```
@taldcroft , is this expected?

pllim commented on astropy/timeseries/io/tests/test_kepler.py:
```diff
@@ -80,7 +80,8 @@ def test_kepler_astropy():
 def test_tess_astropy():
     filename = get_pkg_data_filename('timeseries/hlsp_tess-data-alerts_tess_phot_00025155310-s01_tess_v1_lc.fits')
-    timeseries = kepler_fits_reader(filename)
+    with pytest.warns(UserWarning, match='Ignoring 815 rows with NaN times'):
```
@astrofrog , is this expected?

astrofrog replied:
Yes this is fine, thanks!

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,19 @@ doctest_plus = enabled
+    ignore:The timefunc function is deprecated
+    ignore:numpy.ufunc size changed:RuntimeWarning
+    ignore:Remote data cache could not be accessed due to NotADirectoryError
+    ignore:.*Cache directory cannot be read or created
```
@aarchiba , some remote data tests (both in test files and doctest) still emit these two warnings, so I ignore them for now. We can revisit after 4.0.

bsipocz commented:
try to turn it off and on again? 🤷‍♀

astrofrog commented on astropy/io/fits/tests/test_hdulist.py:
```diff
@@ -620,8 +619,7 @@ def test_update_with_truncated_header(self):
 
-    @pytest.mark.xfail(platform.system() == 'Windows',
-                       reason='https://github.com/astropy/astropy/issues/5797')
+    # https://github.com/astropy/astropy/issues/5797
```
Is the issue link still needed if this is no longer an issue?

pllim replied:
I prefer to leave the comment in, in case it comes back to haunt us, because I have no idea why the `xfail` suddenly pass again in this PR when we are not even trying to fix it.

astrofrog replied:
I'm adding a one-line comment before that to explain why the issue link is there.

astrofrog commented on astropy/io/fits/tests/test_image.py:
```diff
@@ -241,8 +241,7 @@ def test_primary_with_extname(self):
 
-    @pytest.mark.xfail(platform.system() == 'Windows',
-                       reason='https://github.com/astropy/astropy/issues/5797')
+    # https://github.com/astropy/astropy/issues/5797
```
As above

astrofrog commented on docs/timeseries/index.rst:
```diff
@@ -267,8 +270,11 @@ the data to determine the baseline flux::
-    >>> ts_binned = aggregate_downsample(ts_folded, time_bin_size=0.03 * u.day)  # doctest: +REMOTE_DATA
+    >>> with warnings.catch_warnings():
+    ...     warnings.simplefilter('ignore')  # Ignore empty slice warning
+    ...     ts_binned = aggregate_downsample(ts_folded, time_bin_size=0.03 * u.day)  # doctest: +REMOTE_DATA
```
Use ``+IGNORE_WARNINGS`` here too

pllim replied:
You got to it before I did. 😉

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,20 @@ doctest_plus = enabled
+    ignore:numpy.ufunc size changed:RuntimeWarning
+    ignore:Remote data cache could not be accessed due to NotADirectoryError
+    ignore:.*Cache directory cannot be read or created
+    ignore:can't resolve package from __spec__
```
Are you sure we need this, @astrofrog ? Do you see it when you run with `python setup.py test` instead of `pytest` directly? I thought we have suppressed it either in

https://github.com/astropy/astropy/blob/93901211dc3c5074325068604b5b69e0c93e1db2/astropy/tests/helper.py#L131

or

https://github.com/astropy/astropy/blob/93901211dc3c5074325068604b5b69e0c93e1db2/astropy/tests/helper.py#L149-L151

astrofrog replied:
I saw it when running the tests for the docs with ``pytest docs`` which bypasses the stuff in astropy/tests/helper.py. Ideally we should just define all the things to ignore in ``setup.cfg`` now rather than in ``astropy/tests/helper.py``. So can we leave this entry in ``setup.cfg``, and I'll look shortly (after this PR) into whether we can combine some of the stuff in ``helper.py`` with ``setup.cfg``?

pllim replied:
@astrofrog , if you want to take a stab at it, there is also some incantations in `astropy/conftest.py`.

https://github.com/astropy/astropy/blob/93901211dc3c5074325068604b5b69e0c93e1db2/astropy/conftest.py#L25

bsipocz replied:
yes, this should not be in `astropy/astropy/tests/helper.py` as that file will be picked up by other packages, too that use the test runner, and they may or may not want to ignore all of these (in practice they probably want to but you get the point)

pllim replied:
If you remove it now, they are gonna suddenly see all these warnings... 🤷‍♀

bsipocz replied:
my point is that they should rather live in the conftest, shouldn't they?

pllim replied:
I think Tom R wants to move them out to `setup.cfg` altogether.

bsipocz replied:
that's out of the scope I believe, wouldn't that be a new feature after feature freeze? :trollface:

bsipocz replied:
(btw, what's wrong with conftest.py as opposed to setup.cfg?)

astrofrog replied:
Definitely out of scope for this PR! Let's chat about this at the coordination meeting, it's not urgent. I was just trying to justify adding a warning to ignore to ``setup.cfg``. The short answer @bsipocz is that if you want to be able to do things like ``pytest docs/coordinates`` then ``astropy/conftest.py`` is ignored.

pllim commented on astropy/coordinates/tests/test_celestial_transformations.py:
```diff
@@ -101,7 +104,9 @@ def test_galactocentric():
 
     g_xyz = icrs_coord.transform_to(Galactic).cartesian.xyz
-    gc_xyz = icrs_coord.transform_to(Galactocentric(z_sun=0*u.kpc)).cartesian.xyz
+    with galactocentric_frame_defaults.set('pre-v4.0'):
```
This got a deprecation warning but only in 32-bit CircleCI job. 🤷‍♀

pllim commented:
CircleCI 32-bit has the following warning but it is not failing the test suite, so just FYI:
```
.../python3.8/site-packages/_pytest/config/__init__.py:830
  .../config/__init__.py:830: PytestAssertRewriteWarning:
  Module already imported so cannot be rewritten: xdist
    self._mark_plugins_for_rewrite(hook)
```

pllim commented on setup.cfg:
```diff
@@ -100,10 +100,23 @@ doctest_plus = enabled
+    ignore:unclosed <ssl.SSLSocket:ResourceWarning
+    ignore:The timefunc function is deprecated
+    ignore:numpy.ufunc size changed:RuntimeWarning
+    ignore:.*'np.bool_' scalars to be interpreted as an index:DeprecationWarning
```
@astrofrog , is this the correct way to handle this warning that I see in https://travis-ci.org/astropy/astropy/jobs/612683779 coming out of visualization tests?

```
        left, right = self.xaxis.limit_range_for_scale(left, right)
>       left, right = sorted([left, right], reverse=reverse)
E       DeprecationWarning: In future, it will be an error for 'np.bool_' scalars to be interpreted as an index
.../python3.8/site-packages/matplotlib/axes/_base.py:3270: DeprecationWarning
```

astrofrog replied:
I think we can just get the 3.2 pre-release of Matplotlib installed on 3.8 instead - I'll push up a fix.

pllim commented on astropy/units/tests/test_quantity_ufuncs.py:
```diff
@@ -710,8 +712,15 @@ def test_onearg_test_ufuncs(self, ufunc):
         assert np.all(out == ufunc(q.value))
 
     def test_sign(self):
+        if sys.platform.startswith('win'):
```
@mhvk, is this expected?

astrofrog replied:
If so, a cleaner way to catch this would be to use Numpy's ``errstate``:

```
with np.errstate(invalid='ignore'):
    ...
```

which means we don't need to both with conditionally selecting ``nullcontext``.

In general I'd use ``pytest.warns`` if it's a warning we want to see happen, and ``errstate`` or ``filterwarnings`` (as a decorator) if it's a nuisance but unrelated to the test.

pllim replied:
Is it unrelated though? I am not familiar enough with ufunc internals to know. And it is weird this only happens on Windows.

mhvk replied:
I'd guess this happens even for regular ndarray, so I think we should count it as a nuisance - I think the `filterwarnings` may be the easiest - with a comment explaining that happens on windows and is really an upstream issue.

pllim replied:
Since @nden has identified this to be caused by Numpy bug, I retained the Windows-specific context.

pllim commented on astropy/utils/iers/tests/test_iers.py:
```diff
@@ -330,22 +334,10 @@ def test_IERS_B_parameters_loading_into_IERS_Auto():
                      "correctly to IERS Auto".format(name)))
 
 
-# Issue with FTP, rework test into previous one when it's fixed
```
Hmm... It XPASS now. 🤷‍♀

pllim commented on astropy/wcs/tests/test_pickle.py:
```diff
@@ -86,7 +88,14 @@ def test_wcs():
         os.path.join("data", "outside_sky.hdr"), encoding='binary')
 
-    wcs1 = wcs.WCS(header)
+    if sys.platform.startswith('win'):
```
@nden , is this expected?

astrofrog replied:
If this ends up not being desired behavior, I'd suggest changing to a ``pytest.mark.filterwarnings``, and same below.

pllim commented on astropy/wcs/tests/test_wcs.py:
```diff
@@ -1214,7 +1223,14 @@ def test_footprint_contains():
     assert not hasCoord
 
-    hasCoord = test_wcs.footprint_contains(SkyCoord(24, 2, unit='deg'))
+    if sys.platform.startswith('win'):
```
@nden , is this expected?

astrofrog replied:
Here ``np.errstate`` might make more sense (without the platform conditional)

nden replied:
I don't have access to WIN but I wonder if [this](https://github.com/numpy/numpy/issues/8945#issuecomment-550288279) is the cause.

mhvk replied:
Yes, surely that's it!

Given that it is therefore an upstream issue, maybe for now the best is to stick with a windows-specific `xfail`? At least then when it is fixed we'll notice and can remove it.

pllim replied:
Nice find, @nden. Thanks!

pllim commented on astropy/modeling/blackbody.py:
```diff
@@ -55,10 +55,13 @@ class BlackBody1D(Fittable1DModel):
+    >>> import warnings
     >>> from astropy.modeling import models
     >>> from astropy import units as u
-    >>> bb = models.BlackBody1D()  # doctest: +IGNORE_WARNINGS
```
I really don't understand why this failed in `numpy-dev` job. It makes no sense.

pllim commented:
I think I got it all! The two remote data test failures are unrelated to the warnings we are testing for.

astrofrog commented on astropy/units/tests/test_quantity_ufuncs.py:
```diff
@@ -910,17 +919,24 @@ def test_onearg_test_ufuncs_inplace(self, ufunc):
+            ctx = pytest.warns(
+                RuntimeWarning, match='invalid value encountered in sign')
+        else:
+            ctx = nullcontext()
```
Same comment as above about using ``np.errstate``

astrofrog commented on astropy/utils/iers/tests/test_iers.py:
```diff
@@ -302,11 +303,15 @@ def test_simple(self):
-            assert np.allclose(dat.ut1_utc(Time(60000, format='mjd').jd).value, -0.3)
+            # https://github.com/astropy/astropy/issues/9600
+            with pytest.raises(ValueError):
+                assert np.allclose(dat.ut1_utc(Time(60000, format='mjd').jd).value, -0.3)
```
This doesn't seem right (to make it seem like the exception is wanted). I think an xfail on the test would make more sense here?

astrofrog commented on astropy/utils/iers/tests/test_iers.py:
```diff
@@ -330,22 +335,16 @@ def test_IERS_B_parameters_loading_into_IERS_Auto():
+        iersa_tab = iers.IERS_A.open(url, cache=False)
+    except URLError:
+        if os.environ.get('TRAVIS') and url == iers.IERS_A_URL:
+            pytest.xfail('FTP timed out on Travis CI')
```
Just out of curiosity, can we increase the timeout in IERS_A.open?

pllim replied:
@eteq says it works fine from his machine, so there is also a chance that FTP services sometimes blocks Travis CI after certain threshold. I really don't know. 🤷‍♀  I can just revert this whole thing if you want. It is kinda out of scope anyway, as it resulted in xpass/xfail instead of warnings.

astrofrog replied:
Maybe let's indeed revert and deal with this separately. In my experience Travis also sometimes times out because the outside world connections can be slower.

pllim commented:
@astrofrog , I think I have addressed your remaining comments.

pllim commented on astropy/utils/iers/tests/test_iers.py:
```diff
@@ -251,6 +251,7 @@ def test_no_auto_download(self):
         assert type(t) is iers.IERS_B
 
     @pytest.mark.remote_data
+    @pytest.mark.xfail(reason='See issue 9600')
```
I marked this whole test as xfail until #9600 is fixed. Also reverted my other test modifications to this file, @astrofrog .

pllim commented:
Thanks for the review everyone! I rebased and addressed the latest round of comments. Unless new warnings were introduced in the commits over the weekend, hopefully we're good to go now.

bsipocz commented:
Thanks @pllim!

pllim commented:
Whew. That only took 13 months to clear over 2000 warnings on and off! 😉 

It was very much a group effort. Thank you, everyone!

bsipocz commented:
I'm sure at the end it was more than 2000. Thanks for your persistence, it wouldn't have happen without you.

astrofrog commented:
Thank you @pllim for leading the charge!! 👏 👍

mhvk commented:
Thanks indeed, fantastic!!
